flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManagerTest.java

```diff
@@ -0,0 +1,1092 @@
+			.setResourceTracker(resourceTracker)
+			.buildAndStartWithDirectExec(ResourceManagerId.generate(), resourceManagerActions)) {
+
+			slotManager.processResourceRequirements(resourceRequirements);
+
+			final JobID jobId = resourceRequirements.getJobId();
+			assertThat(getTotalResourceCount(resourceTracker.getMissingResources().get(jobId)), is(1));
+		}
+	}
+
+	/**
+	 * Tests that a slot request which can be fulfilled will trigger a slot allocation.
+	 */
+	@Test
+	public void testSlotRequestWithFreeSlot() throws Exception {
+		final ResourceManagerId resourceManagerId = ResourceManagerId.generate();
+		final ResourceID resourceID = ResourceID.generate();
+		final JobID jobId = new JobID();
+		final SlotID slotId = new SlotID(resourceID, 0);
+		final String targetAddress = "localhost";
+		final ResourceProfile resourceProfile = ResourceProfile.fromResources(42.0, 1337);
+
+		final DefaultSlotTracker slotTracker = new DefaultSlotTracker();
+
+		try (DeclarativeSlotManager slotManager = createDeclarativeSlotManagerBuilder()
+			.setSlotTracker(slotTracker)
+			.buildAndStartWithDirectExec(resourceManagerId, new TestingResourceActionsBuilder().build())) {
+
+			final CompletableFuture<Tuple6<SlotID, JobID, AllocationID, ResourceProfile, String, ResourceManagerId>> requestFuture = new CompletableFuture<>();
+			// accept an incoming slot request
+			final TaskExecutorGateway taskExecutorGateway = new TestingTaskExecutorGatewayBuilder()
+				.setRequestSlotFunction(tuple6 -> {
+					requestFuture.complete(Tuple6.of(tuple6.f0, tuple6.f1, tuple6.f2, tuple6.f3, tuple6.f4, tuple6.f5));
+					return CompletableFuture.completedFuture(Acknowledge.get());
+				})
+				.createTestingTaskExecutorGateway();
+
+			final TaskExecutorConnection taskExecutorConnection = new TaskExecutorConnection(resourceID, taskExecutorGateway);
+
+			final SlotStatus slotStatus = new SlotStatus(slotId, resourceProfile);
+			final SlotReport slotReport = new SlotReport(slotStatus);
+
+			slotManager.registerTaskManager(
+				taskExecutorConnection,
+				slotReport);
+
+			ResourceRequirements requirements = ResourceRequirements.create(
+				jobId,
+				targetAddress,
+				Collections.singleton(ResourceRequirement.create(resourceProfile, 1)));
+			slotManager.processResourceRequirements(requirements);
+
+			assertThat(requestFuture.get(), is(equalTo(Tuple6.of(slotId, jobId, requestFuture.get().f2, resourceProfile, targetAddress, resourceManagerId))));
+
+			DeclarativeTaskManagerSlot slot = slotTracker.getSlot(slotId);
+
+			assertEquals("The slot has not been allocated to the expected allocation id.", jobId, slot.getJobId());
+		}
+	}
+
+	/**
+	 * Checks that reducing resource requirements while a slot allocation is in-progress results in the slot being freed
+	 * one the allocation completes.
+	 */
+	@Test
+	@Ignore
+	public void testResourceRequirementReductionDuringAllocation() throws Exception {
```

Comment:
Is this really supported by the current `DeclarativeSlotManager` implementation?

Reply:
No, that's why it I added `@Ignore` initially, but I never got around to deleting it.

Once the requirements are reduced the SlotManager knows that the pending slots exceeds the requirements, but takes no further action. The TaskExecutor will offer the slot to the JM, which may or may not fail. In any case the TM will inform the SM about the resulting state, so we're good in regards to correctness. (And we'd introduce an odd edge-case where the SM actually frees a slot on the TM)